frontend/src/components/worksheets/items/SchemaItem/SchemaItem.js

```diff
@@ -0,0 +1,343 @@
+        if (clear) {
+            this.clearChanges();
+            return;
+        }
+        this.setState({ editing: !this.state.editing });
+        if (save) {
+            this.saveSchema();
+        }
+    };
+
+    clearChanges = () => {
+        this.setState({ rows: [...this.props.item.field_rows], editing: !this.state.editing });
+    };
+
+    saveSchema = () => {
+        const { schema_name, field_rows } = this.props.item;
+        let updatedSchema = [];
+        let fromAddSchema = false;
+        let schemaBlockSourceLength = field_rows.length;
+        this.state.rows.forEach((fields) => {
+            if (!fields.field) {
+                return;
+            }
+            if (!fromAddSchema && fields.from_schema_name !== schema_name) {
+                // these rows correspond to addschema
+                fromAddSchema = true;
+                updatedSchema.push('% addschema ' + fields.from_schema_name);
+                return;
+            } else if (fromAddSchema && fields.from_schema_name !== schema_name) {
+                // These rows doesn't occupy any source lines
+                schemaBlockSourceLength -= 1;
+                return;
+            } else {
+                fromAddSchema = false;
+            }
+
+            let curRow = '% add ' + fields.field;
+            if (!fields['generated-path']) {
+                updatedSchema.push(curRow);
+                return;
+            }
+            curRow = curRow + ' ' + fields['generated-path'];
+            if (!fields['post-processing']) {
+                updatedSchema.push(curRow);
+                return;
+            }
+            curRow = curRow + ' ' + fields['post-processing'];
+            updatedSchema.push(curRow);
+        });
+        this.props.updateSchemaItem(
+            updatedSchema,
+            this.props.item.start_index,
+            schemaBlockSourceLength,
+        );
+    };
+
+    addFieldRowAfter = (idx) => (e) => {
+        if (!this.props.editPermission) return;
+        const schemaItem = this.props.item;
+        const schemaHeaders = schemaItem.header;
+        let newRow = { from_schema_name: schemaItem.schema_name };
+        schemaHeaders.forEach((header) => {
+            newRow[header] = '';
+        });
+        let curRow = this.state.rows;
+        curRow.splice(idx + 1, 0, newRow);
+        this.setState({ rows: curRow });
+    };
+
+    changeFieldValue = (idx, key) => (e) => {
+        if (!this.props.editPermission) return;
+        const { rows } = this.state;
+        let copiedRows = JSON.parse(JSON.stringify(rows));
+        copiedRows[idx][key] = e.target.value;
+        this.setState({ rows: [...copiedRows] });
+    };
+
+    moveFieldRow = (idx, direction) => () => {
+        if (!this.props.editPermission) return;
+        // -1 for moving up, 1 for moving down
+        const { rows } = this.state;
+        let copiedRows = JSON.parse(JSON.stringify(rows));
```

Comment:
I think we can just do this here, because we're only modifying the order of the rows (not the properties of each element in `rows` itself)

```suggestion
        let copiedRows = [...rows];
```